cmd/ingester/app/consumer/rate_limiter.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consumer
+
+import (
+	"math"
+	"time"
+)
+
+const cost = 1.0
+
+type config struct {
+	creditsPerSecond float64
+	maxBalance       float64
+}
+
+type rateLimiter struct {
+	config
+	balance  float64
+	lastTime time.Time
+	event    chan config
+}
+
+// newRateLimiter constructs a new rateLimiter.
+func newRateLimiter(creditsPerSecond, maxBalance float64) *rateLimiter {
+	r := &rateLimiter{
+		config: config{
+			creditsPerSecond: creditsPerSecond,
+			maxBalance:       maxBalance,
+		},
+		balance: maxBalance,
+		event:   make(chan config),
+	}
+	r.lastTime = time.Now()
+	return r
+}
+
+// Acquire causes rateLimiter to spend a credit and return immediately,
+// or block until a credit becomes available. Returns true if credits were
+// acquired. Returns false if Stop was called before credits were acquired,
+// interrupting the Acquire call.
+// Note: Acquire has no corresponding release operation.
+// N.B. This rate limiter implementation is not thread-safe. Do not call Acquire
+// from more than one goroutine.
+func (r *rateLimiter) Acquire() bool {
+	for r.updateBalance() < cost {
+		timer := time.NewTimer(r.calculateWait())
```

Comment:
What is the benefit of having `Acquire` being interruptible? 
Creating a new timer whenever the balance is less than the cost is quite convoluted - an alternative is to sleep for the calculated wait time.

Reply:
Or I could always use `time.AfterFunc`. I'm not sure it makes sense to implement a blocking rate limiter without the ability to cancel/interrupt any blocking calls. Sounds like it might lead to unexpected hangs in an application down the line. Also, this allows us to handle update channels etc. while we sleep.